v0/destinations/personalize/transform.js

```diff
@@ -18,75 +18,93 @@ async function process(event) {
-        !(
-          keyField[k].toUpperCase() === "USER_ID" ||
-          keyField[k].toUpperCase() === "EVENT_TYPE" ||
-          keyField[k].toUpperCase() === "TIMESTAMP" ||
-          keyField[k].toUpperCase() === "ITEM_ID" ||
-          keyField[k].toUpperCase() === "EVENT_VALUE" ||
-          keyField[k].toUpperCase() === "IMPRESSION" ||
-          keyField[k].toUpperCase() === "RECOMMENDATION_ID" ||
-          keyField[k].toUpperCase() === "EVENT_ID"
-        )
-      ) {
-        const mappedFields = mappedField[k];
-        keyField[k] = keyField[k]
-          .toLowerCase()
-          .replace(/_([a-z])/g, function camelCase(g) {
-            return g[1].toUpperCase();
-          });
-        if (properties && properties[mappedFields]) {
-          property[keyField[k]] = properties[mappedFields];
-        } else if (typeof message[mappedFields] !== "undefined") {
-          property[keyField[k]] = message[mappedFields];
+    for (let j = 0; j < noOfFields; j += 1) {
+      mappedField.push(Config.customMappings[j].to);
+    }
+    const mappedEvent = mapEventName(event);
+    if (mappedEvent) {
+      for (let k = 0; k < noOfFields; k += 1) {
+        if (
+          // except these every thing will go inside properties
+          !(
+            keyField[k].toUpperCase() === "USER_ID" ||
+            keyField[k].toUpperCase() === "EVENT_TYPE" ||
+            keyField[k].toUpperCase() === "TIMESTAMP" ||
+            keyField[k].toUpperCase() === "ITEM_ID" ||
+            keyField[k].toUpperCase() === "EVENT_VALUE" ||
+            keyField[k].toUpperCase() === "IMPRESSION" ||
+            keyField[k].toUpperCase() === "RECOMMENDATION_ID" ||
+            keyField[k].toUpperCase() === "EVENT_ID"
+          )
+        ) {
+          const mappedFields = mappedField[k];
+          keyField[k] = keyField[k]
+            .toLowerCase()
+            .replace(/_([a-z])/g, function camelCase(g) {
+              return g[1].toUpperCase();
+            });
+          if (properties && properties[mappedFields]) {
+            property[keyField[k]] = properties[mappedFields];
+          } else {
+            throw new Error(`Mapped Field ${mappedFields} not found`);
+          }
+        } else if (
+          // userId and eventType is handled later as shown below and others are saved outside properties
+          !(
+            keyField[k].toUpperCase() === "USER_ID" ||
+            keyField[k].toUpperCase() === "EVENT_TYPE"
+          )
+        ) {
+          keyField[k] = keyField[k]
+            .toLowerCase()
+            .replace(/_([a-z])/g, function camelCase(g) {
+              return g[1].toUpperCase();
+            });
+          const mappedFields = mappedField[k];
+          eventObj[keyField[k]] = properties[mappedFields];
+        }
+      }
+      // itemId is a mandatory field, so even if user doesn't mention, it is needed to be provided
+
+      if (properties.itemId) {
```

Comment:
why are you checking properties.itemId? whatever mapped value for item id is there it should be searched in properties and set. 

ITEM_ID ---> VALUE1

check if value1 is present in property. If not then set to messageId. 
same logic for userId